PsfLauncher/main.cpp

```diff
@@ -268,43 +276,278 @@ void LaunchMonitorInBackground(std::filesystem::path packageRoot, const wchar_t
+
+        if (waitResult != WAIT_OBJECT_0)
+        {
+            auto err{ ::GetLastError() };
+            return { L"Waiting for the application terminated unexpectedly.", err };
+        }
+    }
+    else
+    {
+        std::wostringstream ss;
+        auto err{ ::GetLastError() };
+        ss << L"ERROR: Failed to create a process for " << GetApplicationNameFromExecInfo(execInfo);
+
+        return { ss.str(), err };
+    }
+
+    DWORD exitCode = ERROR_SUCCESS;
+    if (GetExitCodeProcess(processInfo.hProcess, &exitCode))
+    {
+        if (exitCode != ERROR_SUCCESS)
+        {
+            return { L"ERROR: Application failed for " + GetApplicationNameFromExecInfo(execInfo), exitCode };
+        }
+    }
+    else
+    {
+        return { L"Could not get the exit code process for " + GetApplicationNameFromExecInfo(execInfo), ::GetLastError() };
+    }
+
+    return {};
+}
+
+std::wstring GetApplicationNameFromExecInfo(ExecutionInformation execInfo)
+{
+    if (execInfo.ApplicationName)
+    {
+        return execInfo.ApplicationName;
+    }
+    else
+    {
+        std::wstring applicationName;
+        //If the application name contains spaces, the application needs to be surrounded in quotes.
+        if (execInfo.CommandLine[0] == '"')
+        {
+            //Skip the first quote and don't include the last quote.
+            applicationName = execInfo.CommandLine.substr(1, execInfo.CommandLine.find('"', 1) - 1);
         }
         else
         {
-            if (wait)
-                WaitForSingleObject(processInfo.hProcess, INFINITE);
+            applicationName = execInfo.CommandLine.substr(0, execInfo.CommandLine.find(' '));
+
         }
+
+        return applicationName;
     }
 }
-int __stdcall wWinMain(HINSTANCE, HINSTANCE, PWSTR args, int cmdShow)
+
+ErrorInformation StartWithShellExecute(std::filesystem::path packageRoot, std::filesystem::path exeName, std::wstring exeArgString, LPCWSTR dirStr, int cmdShow) noexcept
 {
-    return launcher_main(args, cmdShow);
+    // Non Exe case, use shell launching to pick up local FTA
+    auto nonExePath = packageRoot / exeName;
+
+    SHELLEXECUTEINFO shex = {
+        sizeof(shex)
+        , SEE_MASK_NOCLOSEPROCESS
+        , (HWND)nullptr
+        , nullptr
+        , nonExePath.c_str()
+        , exeArgString.c_str()
+        , dirStr ? (packageRoot / dirStr).c_str() : nullptr
+        , static_cast<WORD>(cmdShow)
+    };
+
+
+    Log("\tUsing Shell launch: %ls %ls", shex.lpFile, shex.lpParameters);
+    if (!ShellExecuteEx(&shex))
+    {
+        auto err{ ::GetLastError() };
+        return { L"ERROR: Failed to create detoured shell process", err };
+    }
+    
+    DWORD exitCode = ERROR_SUCCESS;
+    if (GetExitCodeProcess(shex.hProcess, &exitCode))
+    {
+        if (exitCode != ERROR_SUCCESS)
+        {
+            return { L"Application exited and was not successful. " + exeName.native(), exitCode };

```

Comment:
consider factoring out a function for retrieving/reporting exit code, in part to standardize on message text